lahja/endpoint.py

```diff
@@ -61,36 +64,52 @@ def from_name(cls, name: str, base_path: Optional[pathlib.Path]=None) -> 'Connec
 
 
-class EndpointConnector:
-    """
-    Expose the receiving queue of an :class:`~lahja.endpoint.Endpoint` so that any other
-    :class:`~lahja.endpoint.Endpoint` that wants to push events into this
-    :class:`~lahja.endpoint.Endpoint` can do so via the
-    :class:`~lahja.endpoint.ProxyEndpointConnector`.
-    """
+class Broadcast(NamedTuple):
+    event: Union[BaseEvent, bytes]
+    config: Optional[BroadcastConfig]
 
-    def __init__(self, endpoint: 'Endpoint') -> None:
-        self._endpoint = endpoint
 
-    def put_nowait(self, item_and_config: Tuple[BaseEvent, Optional[BroadcastConfig]]) -> None:
-        loop = self._endpoint._loop
-        if not self._endpoint._running:
-            self._endpoint.logger.warning(
-                "Attempted to push into %s while it isn't running.",
-                self._endpoint.name
-            )
-        # We need to wrap this in `call_soon_threadsafe` since otherwise, the event loop
-        # won't pick it up until some other task moves the loop forward
-        loop.call_soon_threadsafe(self._endpoint._receiving_queue.put_nowait, item_and_config)
+class RemoteEndpoint:
+    def __init__(self, reader: StreamReader, writer: StreamWriter) -> None:
+        self.writer = writer
+        self.reader = reader
+        self._drain_lock = asyncio.Lock()
 
+    @staticmethod
+    async def connect_to(path: str) -> 'RemoteEndpoint':
+        reader, writer = await asyncio.open_unix_connection(path)
+        return RemoteEndpoint(reader, writer)
 
-class ProxyEndpointConnector(BaseProxy):  # type: ignore # TypeShed is missing BaseProxy
-    """
-    Proxy that connects to an :class:`~lahja.endpoint.EndpointConnector`
-    """
+    async def send_message(self, message: Broadcast) -> None:
+        await _write_message(self.writer, message)
+        async with self._drain_lock:
+            # Use a lock to serialize drain() calls. Circumvents this bug:
+            # https://bugs.python.org/issue29930
+            await self.writer.drain()
 
-    def put_nowait(self, item_and_config: Tuple[BaseEvent, Optional[BroadcastConfig]]) -> None:
-        self._callmethod('put_nowait', (item_and_config,))
+    async def read_message(self) -> Broadcast:
+        return await _read_message(self.reader)
+
+    @property
+    def more_to_read(self) -> bool:
+        return not self.reader.at_eof()
+
+
+async def _read_message(reader: StreamReader) -> Broadcast:
+    raw_size = await reader.readexactly(4)
+    size = int.from_bytes(raw_size, 'little')
+    message = await reader.readexactly(size)
+    obj = pickle.loads(message)
+    assert isinstance(obj, Broadcast)  # maybe this should be a user-friendly error?
+    return obj
+
+
+async def _write_message(writer: StreamWriter, message: Broadcast) -> None:
+    assert isinstance(message, Broadcast)
+    pickled = pickle.dumps(message)
+    size = len(pickled)
+    writer.write(size.to_bytes(4, 'little'))
```

Comment:
So this would refer to `SIZE_MARKER_LENGTH` (for the lack of a better name) as well